v2/pkg/engine/resolve/resolve_test.go

```diff
@@ -1876,6 +1876,64 @@ func TestResolver_ResolveGraphQLResponse(t *testing.T) {
 			},
 		}, Context{ctx: context.Background()}, `{"data":{"user":{"id":1,"name":"Jannik","__typename":"User","aliased":"User","rewritten":"User"}}}`
 	}))
+	t.Run("__typename checks", testFn(func(t *testing.T, ctrl *gomock.Controller) (node *GraphQLResponse, ctx Context, expectedOutput string) {
+		return &GraphQLResponse{
+			Fetches: Single(&SingleFetch{
+				FetchConfiguration: FetchConfiguration{DataSource: FakeDataSource(`{"id":1,"name":"Jannik","__typename":"NotUser","rewritten":"User"}`)},
+			}),
+			Data: &Object{
+				Fields: []*Field{
+					{
+						Name: []byte("user"),
+						Value: &Object{
+							Fields: []*Field{
+								{
+									Name: []byte("id"),
+									Value: &Integer{
+										Path:     []string{"id"},
+										Nullable: false,
+									},
+								},
+								{
+									Name: []byte("name"),
+									Value: &String{
+										Path:     []string{"name"},
+										Nullable: false,
+									},
+								},
+								{
+									Name: []byte("__typename"),
+									Value: &String{
+										Path:          []string{"__typename"},
+										Nullable:      false,
+										IsTypeName:    true,
+										AllowedValues: map[string]struct{}{"User": {}},
+										SourceName:    "Users",
+									},
+								},
+								{
+									Name: []byte("aliased"),
+									Value: &String{
+										Path:       []string{"__typename"},
+										Nullable:   false,
+										IsTypeName: true,
+									},
+								},
+								{
+									Name: []byte("rewritten"),
+									Value: &String{
+										Path:       []string{"rewritten"},
+										Nullable:   false,
+										IsTypeName: true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}, Context{ctx: context.Background()}, `{"errors":[{"message":"Subgraph 'Users' returned invalid value 'NotUser' for __typename field.","path":["__typename"]}],"data":null}`
```

Comment:
How does it behave with different error formats? Does it respect the propagation settings? By default it should be wrapped but this looks like passthrough. It looks to me that we handle errors differently here in the engine.

Reply:
This is not an error originating from a Subgraph. A subgraph is misbehaving and returning invalid data for a field. The engine (Router) catches it and ensures that the client doesn't rely on incorrect information by bubbling up an error, setting the field to nil, and propagating an extensions code.

Reply:
My question: What is responded to the client? If we propagate this error in all subgraph propagation modes, we have inconsistencies and this will break clients.